ledger/complete/wal/checkpoint_v6_writer.go

```diff
@@ -0,0 +1,529 @@
+package wal
+
+import (
+	"bufio"
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/ledger/complete/mtrie/flattener"
+	"github.com/onflow/flow-go/ledger/complete/mtrie/node"
+	"github.com/onflow/flow-go/ledger/complete/mtrie/trie"
+	utilsio "github.com/onflow/flow-go/utils/io"
+)
+
+const subtrieLevel = 4
+const subtrieCount = 1 << subtrieLevel
+
+type NodeEncoder func(node *trie.MTrie, index uint64, scratch []byte) []byte
+
+type resultStoringSubTrie struct {
+	Index     int
+	Roots     map[*node.Node]uint64 // node index for root nodes
+	NodeCount uint64
+	Checksum  uint32
+	Err       error
+}
+
+// StoreCheckpointV6 stores checkpoint file into a main file and 17 file parts.
+// the main file stores:
+// 		1. version
+//		2. checksum of each part file (17 in total)
+// 		3. checksum of the main file itself
+// 	the first 16 files parts contain the trie nodes below the subtrieLevel
+//	the last part file contains the top level trie nodes above the subtrieLevel and all the trie root nodes.
+func StoreCheckpointV6(
+	tries []*trie.MTrie, outputDir string, outputFile string, logger *zerolog.Logger) error {
+	if len(tries) == 0 {
+		logger.Info().Msgf("no tries to be checkpointed")
+		return nil
+	}
+
+	first, last := tries[0], tries[len(tries)-1]
+	logger.Info().
+		Str("first_hash", first.RootHash().String()).
+		Uint64("first_reg_count", first.AllocatedRegCount()).
+		Str("last", last.RootHash().String()).
+		Uint64("last_reg_count", last.AllocatedRegCount()).
+		Msgf("storing checkpoint for %v tries to %v", len(tries), outputDir)
+
+	subtrieRoots := createSubTrieRoots(tries)
+
+	estimatedSubtrieNodeCount := estimateSubtrieNodeCount(tries)
+
+	subTrieRootIndices, subTriesNodeCount, subTrieChecksums, err := storeSubTrieConcurrently(
+		subtrieRoots,
+		estimatedSubtrieNodeCount,
+		outputDir,
+		outputFile,
+		logger,
+	)
+	if err != nil {
+		return fmt.Errorf("could not store sub trie: %w", err)
+	}
+
+	logger.Info().Msgf("subtrie have been stored. sub trie node count: %v", subTriesNodeCount)
+
+	topTrieChecksum, err := storeTopLevelNodesAndTrieRoots(
+		tries, subTrieRootIndices, subTriesNodeCount, outputDir, outputFile, logger)
+	if err != nil {
+		return fmt.Errorf("could not store top level tries: %w", err)
+	}
+
+	err = storeCheckpointSummary(subTrieChecksums, topTrieChecksum, outputDir, outputFile, logger)
+	if err != nil {
+		return fmt.Errorf("could not store checkpoint summary: %w", err)
+	}
+
+	logger.Info().Msgf("checkpoint file has been successfully stored at %v/%v", outputDir, outputFile)
+
+	return nil
+}
+
+// 		1. version
+//		2. checksum of each part file (17 in total)
+// 		3. checksum of the main file itself
+func storeCheckpointSummary(
```

Comment:
Does this file need to have a custom binary format?  Maybe we can just store a json here?

Reply:
In V5, there is only a single file named "root.checkpoint", for instance, and this summary file in V6 is essentially the same file as that. The name summary is not the best name, I changed it into header in another PR.

Reply:
My question is mostly why do we need to deal with all that custom binary encoding/decoding instead of just creating a struct that specifies a "summary" of the checkpoint and just `Marshal`/`Unmarshal` it in any format that we want.  Is custom marshaling/unmarshaling code worth it?

Reply:
Track this idea here for now https://github.com/dapperlabs/flow-go/issues/6403